### PR TITLE
Focal loss

### LINKED ISSUE #115: Port focal loss
a subtask of #84, port a focal loss function into Monai

e.g.:
https://arxiv.org/abs/1708.02002

### PR DESCRIPTION
**Pull request by LucasFidon**
Fixes #115 Port focal loss .

### Description
Add an implementation of the Focal loss.

### Status
**Ready**

### Types of changes
- [x] New functionality added
- [x] Added unit tests for the new functionality

### CODE REVIEW

File: monai/losses/focal_loss.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+        # Get the proba
+        pt = torch.exp(logpt)  # N,H*W
+
+        if self.alpha is not None:
+            if self.alpha.type() != i.data.type():
+                self.alpha = self.alpha.type_as(i.data)
+            # Select the correct weight for each voxel depending on its
+            # associated gt label
+            at = torch.unsqueeze(self.alpha, dim=0)  # C => 1,C
```
Comment (ericspod) [inline monai/losses/focal_loss.py:77]: Pytorch allows you to use None as an index value to unsqueeze, so lines 77-8 can be `at = at[None, :, None]`. Similarly squeezing can be done by just indexing a dimension, line 81 becomes `at = at[:, 0]`. It's a style choice but does reduce the code a bit if it helps and doesn't impact readability.

File: monai/losses/focal_loss.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            at = torch.unsqueeze(at, dim=2)  # 1,C => 1,C,1
+            at = at.expand((t.size(0), -1, t.size(2)))  # 1,C,1 => N,C,H*W
+            at = at.gather(1, t.data)  # selection of the weights  => N,1,H*W
+            at = torch.squeeze(at, dim=1)  # N,1,H*W => N,H*W
+            # Multiply the log proba by their weights
+            logpt = logpt * Variable(at)
```
Comment (ericspod) [inline monai/losses/focal_loss.py:83]: The `Variable` API is [deprecated](https://pytorch.org/docs/stable/autograd.html#variable-deprecated). This will work just fine but eventually `Variable` will disappear.

File: monai/losses/focal_loss.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    """
+    PyTorch implementation of the Focal Loss.
+    [1] "Focal Loss for Dense Object Detection", T. Lin et al., ICCV 2017
+    """
+    def __init__(self, gamma=2., alpha=None, reduction='mean'):
+        """
+        Args:
+            gamma: (float) value of the exponent gamma in the definition
+            of the Focal loss.
+            alpha: (float or float list or None) weights to apply to the
```
Comment (ericspod) [inline monai/losses/focal_loss.py:28]: I would suggest describing alpha in a little more detail, what shape should it have it not a single float? If the loss is being used as a categorical loss for a classifier, there's the NxC prediction tensor for N instances with C classes, should alpha then be shape N? If I want to weight each class then do I need to compute alpha for each minibatch, assigning the weight corresponding to the ground truth class assignment at each instance N?

Comment (LucasFidon) [reply monai/losses/focal_loss.py:28]: Thank you @ericspod for taking the time to review the code!

In the Focal loss paper alpha is static (it does not depend on the iteration or the mini-batch) and is of shape C. 

In fact, alpha is exactly the same as the weights of a weighted cross entropy. So it is probably best to align the API with the one of the cross entropy loss of PyTorch https://pytorch.org/docs/stable/_modules/torch/nn/modules/loss.html#CrossEntropyLoss
i.e. the Focal loss would inherit from _WeightedLoss and alpha would be replaced by the parameter weight (which will be a tensor).

Please let me know what you think about this.

Comment (ericspod) [reply monai/losses/focal_loss.py:28]: I understand a bit better now what the implementation here is doing versus [this one](https://github.com/clcarwin/focal_loss_pytorch/blob/master/focalloss.py), and is doing the same thing. An example of using the loss would help (which we need elsewhere as well):

```python
pred=torch.tensor([[1,0],[0,1],[1,0]],dtype=torch.float32)
grnd=torch.tensor([0,1,0],dtype=torch.int64)
fl=FocalLoss()
fl(pred,grnd)
``` 

The `_WeightedLoss` type doesn't add much except the `weight` attribute so it's probably a good idea. In that case as well the `reduction` argument of the `_Loss` constructor should be used as well to set the attribute `reduction` as expected rather than overwriting it.

File: monai/losses/focal_loss.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            t = t.unsqueeze(1)  # N,H,W => N,1,H,W
+        if input.dim() > 2:
+            i = i.view(i.size(0), i.size(1), -1)  # N,C,H,W => N,C,H*W
+            t = t.view(t.size(0), t.size(1), -1)  # N,1,H,W => N,1,H*W
+        else:  # Compatibility with classification
+            i = i.unsqueeze(2)  # N,C => N,C,1
+            t = t.unsqueeze(2)  # N,1 => N,1,1
+
+        # Compute the log proba (more stable numerically than softmax)
+        logpt = F.log_softmax(i, dim=1)  # N,C,H*W
```
Comment (wyli) [inline monai/losses/focal_loss.py:64]: came across sigmoidfocal and softmaxfocal in the pytorch repo (for 2D):
https://github.com/pytorch/pytorch/blob/821b5f138a987807032a2fd908fe10a5be5439d9/modules/detectron/sigmoid_focal_loss_op.cu#L26
https://github.com/pytorch/pytorch/blob/821b5f138a987807032a2fd908fe10a5be5439d9/modules/detectron/softmax_focal_loss_op.cu#L59
shall we consider both options here?

ref https://github.com/clcarwin/focal_loss_pytorch/issues/7

Comment (ericspod) [reply monai/losses/focal_loss.py:64]: The sigmoid formulation is for binary classification I think.
